bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java

```diff
@@ -304,6 +304,36 @@ public void resumeMinorGC() {
         }
     }
 
+
+    /**
+     * gc ledger storage.
+     */
+    void gc() {
+        LOG.info("Starting garbage collecting ledger storage.");
+
+        LOG.info("Garbage collecting deleted ledgers' index files.");
+        // gc inactive/deleted ledgers
+        doGcLedgers();
+
+        LOG.info("Scanning entry log files to extract metadata and delete empty entry logs if possible.");
+        // Extract all of the ledger ID's that comprise all of the entry logs
+        // (except for the current new one which is still being written to).
+        entryLogMetaMap = extractMetaFromEntryLogs(entryLogMetaMap);
+
+        // if it isn't running, break to not access zookeeper
+        if (!running) {
+            return;
+        }
+
+        LOG.info("Garbage collecting deleted ledgers' index files again just in case ledgers deleted during scanning.");
+        // gc inactive/deleted ledgers again, just in case ledgers are deleted during scanning entry logs
+        doGcLedgers();
+
+        LOG.info("Garbage collecting empty entry log files.");
+        // gc entry logs
+        doGcEntryLogs();
```

Comment:
In the case of your disk is full, the bookie cannot replay the journal. so the bookie can't actually start to reclaim the disk. The logic below is to ensure bookie can still do gc when the disks are full.

`
        try {
            checkDiskSpace();
        } catch (NoWritableLedgerDirException nwlde) {
            LOG.info("Ledger storage is already full : ", nwlde);
            // if there is no writable ledger dir, we should try reclaimSpace
            reclaimDiskSpace();
            // check disk again if we still can't have enough room for replaying journal, fail it
            checkDiskSpace();
        }
`